File: basicqueue.java
import java.util.Scanner;

public class basicqueue {
    private int[] array;
    private int front;
    private int rear;
    private static int max_size=100;
  

    public basicqueue(int size) {
        array = new int[size];
        front = -1;
        rear = -1;
    }

    public boolean isEmpty() {
        return front == -1;
    }

    public boolean isFull() {
        return rear == array.length - 1;
    }

    public void enqueue(int value) {
        if (isFull()) {
            System.out.println("Queue is full. Cannot enqueue.");
            return;
        }
        if (isEmpty()) {
            front = 0;
        }
        rear++;
        array[rear] = value;
    }

    public int dequeue() {
        if (isEmpty()) {
            System.out.println("Queue is empty. Cannot dequeue.");
            return -1; // Return a sentinel value to indicate an error
        }
        int frontValue = array[front];
        if (front == rear && front != -1 && rear != -1) {
            System.out.println("This element [" + frontValue + "] is dequeued");
        }
        if (front == rear) {
            front = -1;
            rear = -1;
        } else {
            front++;
            System.out.println("This element [" + frontValue + "] is dequeued");
        }
        return frontValue;
    }

    public void display() {
        if (isEmpty()) {
            System.out.println("Queue is empty.");
        } else {
            System.out.print("Queue elements: ");
            for (int i = front; i <= rear; i++) {
                System.out.print(array[i] + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the size of the queue: ");
        int size = scanner.nextInt();

        if (size <= 0 || size>max_size ) {
            System.out.printf("Invalid queue size. Please choose a size between 1 and max size");
            return;
        }

        basicqueue queue = new basicqueue(size);

        int choice;

        while (true) {
            System.out.println("\nQueue Operations:");
            System.out.println("1. Enqueue");
            System.out.println("2. Dequeue");
            System.out.println("3. Display");
            System.out.println("4. Quit");
            System.out.print("Enter your choice: ");
            choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    if (queue.isFull()) {
                        System.out.println("Queue is full. Cannot enqueue.");
                    } else {
                        System.out.print("Enter the item to enqueue: ");
                        int item = scanner.nextInt();
                        queue.enqueue(item);
                    }
                    break;
                case 2:
                    queue.dequeue();
                    break;
                case 3:
                    queue.display();
                    break;
                case 4:
                    System.out.println("Exiting the program.");
                    return;
                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }
    }
}

File: basicqueue.c
#include <stdio.h>
#include <stdbool.h>

#define MAX_SIZE 100

// Structure to represent a queue
struct Queue {
    int array[MAX_SIZE];
    int front, rear;
    int sizes;
};
int main() {
    struct Queue queue;
    initializeQueue(&queue);

     int size;
    printf("Enter the size of the queue: ");
    scanf("%d", &size);
    
    if (size <= 0 || size > MAX_SIZE) {
        printf("Invalid queue size. Please choose a size between 1 and %d.\n", MAX_SIZE);
        return 1;
    }
    
    int choice;
    
    while (1) {
        printf("\nQueue Operations:\n");
        printf("1. enqueue\n");
        printf("2. dequeue\n");
        printf("3. Display\n");
        printf("4. Quit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        
        switch (choice) {
            case 1:
                if (queue.rear == size - 1) {
                    printf("Queue is full. Cannot enqueue.\n");
                } else {
                    int item;
                    printf("Enter the item to enqueue: ");
                    scanf("%d", &item);
                      enqueue(&queue, item);
                }
                break;
            case 2:

                dequeue(&queue);
                break;
            case 3:
                display(queue);
                break;
            case 4:
                printf("Exiting the program.\n");
                return 0;
            
            default:
                printf("Invalid choice. Please try again.\n");
        }
    }


    return 0;
}

// Function to initialize an empty queue
void initializeQueue(struct Queue *queue) {
    queue->front = -1;
    queue->rear = -1;
}



// Function to check if queue is empty
bool isEmpty(struct Queue *queue) {
    return (queue->front == -1);
}



// Function to check if queue is full
bool isFull(struct Queue *queue) {
    return (queue->rear == MAX_SIZE - 1);
}



// Function enqueue
void enqueue(struct Queue *queue, int value) {

    if (isEmpty(queue)) {
        queue->front = 0;
    }
    queue->rear++;
    queue->array[queue->rear] = value;
}



// Function to dequeue
void dequeue(struct Queue *queue) {
       if (isEmpty(queue)) {
        printf("Queue is empty. Cannot dequeue.\n");
    }
    int frontValue = queue->array[queue->front];

       if (queue->front == queue->rear && queue->front != -1 && queue->rear!=-1) {

        printf("this element [%d] is dequed" ,frontValue);
        
        }
    if (queue->front == queue->rear) {
        // If there was only one element in the queue, reset front and rear
        queue->front = -1;
        queue->rear = -1;
    } else {
        queue->front++;
         printf("this element [%d] is dequed" ,frontValue);
    }
     printf("\n");
}

//Function to display
void display(struct Queue queue) {
    if (isEmpty(&queue)) {
        printf("Queue is empty.\n");
    } else {
        printf("Queue elements: ");
        for (int i = queue.front; i <= queue.rear; i++) {
            printf("%d ", queue.array[i]);
        }
        printf("\n");
    }
}










